### ISSUE TITLE
The Dreaded MissingPluginException (Let's have an FAQ)

### ISSUE THREAD

**Issue by ryanheise** [flutter/flutter#62006]
As a plugin developer, I get numerous issues regarding this error on a regular basis. There are many unrelated causes for this error, and to make matters worse, the error message identifies the plugin that was attempting to use platform channels although that is not necessarily the plugin that is responsible for the error, and so you get issues posted by app developers on the wrong projects asking for the bug to be fixed. Furthermore, sometimes the error is caused by an incorrectly written plugin and other times it is caused by something the app developer did or didn't do. In both cases, there is a lack of documentation to help both the plugin developer and the app developer to navigate this issue and track down the root of what's causing it.

What I think is sorely needed is an official MissingPluginException FAQ listing all of the causes for this error and how to address them.

Additionally, since a lot of these errors can be caused whenever an app is creating a background FlutterEngine and it fails to register plugins for some reason, the documentation for Flutter background execution needs to be improved. The following page, for example, still refers to an old Medium article written by Ben Konyi that desperately needs an update:

https://flutter.dev/docs/development/packages-and-plugins/background-processes

On the Android side, the linked article has still not been updated to explain the v2 architecture and plugins that don't do this correctly may fail to do plugin registration correctly in certain circumstances. Yet, this is what the official Flutter documentation refers to.

Here is something that I (as a plugin developer) cannot figure out from the documentation: In that medium article, Ben uses this callback trick on Android to allow plugins to be registered in the background, but it is not needed anymore under v2:

```java
  // In a custom Application subclass
  public void onCreate() {
    super.onCreate();
    GeofencingService.setPluginRegistrant(this);
  }
```

But the same article uses this trick in the iOS implementation:

```objc
- (BOOL)application:(UIApplication *)application
    didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
  // Register the plugins with the AppDelegate
  registerPlugins(self);
  // Set registerPlugins as a callback within GeofencingPlugin. This allows
  // for the Geofencing plugin to register the plugins with the background
  // FlutterEngine instance created to handle events. If this step is skipped,
  // other plugins will not work in the geofencing callbacks!
  [GeofencingPlugin setPluginRegistrantCallback:registerPlugins];

  // Override point for customization after application launch.
  return [super application:application didFinishLaunchingWithOptions:launchOptions];
}
```

Is this trick still required in iOS but not required in Android? If I view the latest version of the geofencing plugin on GitHub, it still has this trick in it. However, the [flutter_isolate](https://pub.dev/packages/flutter_isolate) plugin doesn't do this trick and STILL seems to work in almost most cases, although still people sometimes report that plugins do not get registered in some cases, and it is never obvious what causes it to work with some combination of plugin dependencies and not others. It is typically the presence of one culprit plugin that if removed from the project, allows everything to work again.

In almost every case, it is extremely difficult to track down which plugin is causing plugin registration to fail except by removing plugins one by one until things work again. Perhaps this is also cause to try to improve the error messages so that the real source of the problem can be more easily pinpointed.

**Comment by ryanheise** [comment]
I'm just surveying how various plugins handle the registration of plugins in the background on iOS, and there seems to be some variety and probably not all plugins are doing it correctly.

[FlutterGeofencing](https://github.com/bkonyi/FlutterGeofencing/issues) as mentioned above sets a plugin registrant callback. This callback uses the app's generated plugin registrant to register all plugins. This is the official example linked from the documentation, but to me, this seems wrong if the plugins were already registered automatically because you would run into a "Duplicate plugin error". Perhaps the generated plugin registrant should really check if the plugins are already registered before trying to register them again.

[flutter_downloader](ttps://github.com/fluttercommunity/flutter_downloader) also copies the approach of FlutterGeofencing except that in the callback doesn't use the app's generated plugin registrant, it only attempts to register one plugin and only if it hasn't already been loaded:

```objc
void registerPlugins(NSObject<FlutterPluginRegistry>* registry) {
    if (![registry hasPlugin:@"FlutterDownloaderPlugin"]) {
        [FlutterDownloaderPlugin registerWithRegistrar:[registry registrarForPlugin:@"FlutterDownloaderPlugin"]];
    }
}
```

Putting aside that this is not a general solution as it only loads one plugin, it does appear to be a safer approach to test if the plugin is already registered before attempting to register it. Maybe the official generated plugin registrant should do this for all plugins to prevent a Duplicate plugin error. However that being said, it seems that this if condition has no effect because there is an open issue with `flutter_downloader` suggesting that some users are still getting a Duplicate plugin error: https://github.com/fluttercommunity/flutter_downloader/issues/177 . So it would seem that `[registry hasPlugin:@"FlutterDownloaderPlugin"]` is always returning false/NO and then the following attempt to register still gives a duplicate plugin error. The problem for the plugin developer here is if they remove this code, then it will work fine for users where the plugins happen to be automatically registered, but it will end up with a MissingPluginException where automatic plugin registration fails. It is not clear to me why automatic plugin registration would happen sometimes and not others.

[flutter_isolate](https://github.com/rmawatson/flutter_isolate) just relies on automatic plugin registration. It does not require the plugin user to modify their AppDelegate to set a callback that manually registeres the plugins. This seems to work just fine, although I have heard for some people this does not work and they are getting MissingPluginExceptions.

[audio_service](https://github.com/ryanheise/audio_service) (my plugin) uses flutter_isolate under the hood to manage the background FlutterEngine on iOS and so it comes with the same issues.


Could it be that as the generated plugin registrant goes through each plugin in alphabetical order, if encounters one where `registerWithRegistry` fails, then all of the ones before it will be registered, but all of the ones after it will not?

**Comment by jamesncl** [comment]
There are so many related issues, I don't understand why this isn't being noticed by the Flutter team as a fundamental problem with how plugins are registered. Here are just a handful of currently open picked at random. They all have the same general description - hard to reproduce, it sometimes happens and sometimes magically fixes itself, with no explanation as to why or how. There are literally [hundreds more MissingPluginException issues](https://www.google.com/search?q=missingpluginexception+site%3Ahttps%3A%2F%2Fgithub.com), most of which have been closed, not because it's fixed but because it's not reliably reproducible.

[https://github.com/FirebaseExtended/flutterfire/issues/2675](https://github.com/FirebaseExtended/flutterfire/issues/2675)
[https://github.com/mogol/flutter_secure_storage/issues/71](https://github.com/mogol/flutter_secure_storage/issues/71)
[https://github.com/flutter/flutter/issues/46679](https://github.com/flutter/flutter/issues/46679)
[https://github.com/jonataslaw/VideoCompress/issues/23](https://github.com/jonataslaw/VideoCompress/issues/23)
[https://github.com/flutter/flutter/issues/57443](https://github.com/flutter/flutter/issues/57443)
[https://github.com/pauldemarco/flutter_blue/issues/532](https://github.com/pauldemarco/flutter_blue/issues/532)
[https://github.com/kmcgill88/admob_flutter/issues/21](https://github.com/kmcgill88/admob_flutter/issues/21)
[https://github.com/flutter/flutter/issues/66833](https://github.com/flutter/flutter/issues/66833)
[https://github.com/tekartik/sqflite/issues/452](https://github.com/tekartik/sqflite/issues/452)

**Comment by enyo** [comment]
Just wanted to say, that I have provided an [example that's easy to reproduce](https://github.com/flutter/flutter/issues/66833#issuecomment-703605738) with the most basic setup.

**Comment by blasten** [comment]
Sound like a real pain for someone maintaining a plugin. 

/cc @bparrishMines for your plugin author DevX effort.

**Comment by karora** [comment]
This has been biting me a lot for the last few days, and there seem to be a number of workarounds such as adding `--no-shrink` or downgrading to android-gradle 3.5.  Neither of these are particularly palatable choices, in my own case downgrading to the older android-gradle broke one set of plugins, upgrading to android-gradle 4.1 broke a different set of plugins, and using android-gradle 4.0.2 with `--no-shrink` is where I feel I have force-landed.

Is there a bug somewhere for "switch android templates to use newer android-gradle" because that bug really needs to be aware of this problem. There was [a pull request to do a similar thing back in December 2017](https://github.com/flutter/flutter/pull/13492), and it appears that something like that is needed again.

**Comment by escamoteur** [comment]
It would be interesting to know if this problems only appear in the release builds
If so then it's probably because Proguard strips out the code.

**Comment by MrCsabaToth** [comment]
> It would be interesting to know if this problems only appear in the release builds
> If so then it's probably because Proguard strips out the code.

After I stumbled upon this problem the third time I realized 100% that it's not plugin X or plugin Y's problem. In all of the three times I had this it was release only which makes it just much more painful to debug. Something has to be done, because this will wear down both the app developers and the plugin developers.
So far my experience with Flutter is amazing, but this particular issue is a pressing one.
Just one to add to @jamesncl 's list: https://github.com/miguelpruivo/flutter_file_picker/issues/545
I can dig for the other two occasions. You can see in the issue that others are stumbling upon that, the plugin developer cannot help, the app developers like me try to poke this from all angles... it's just a carnage

**Comment by MrCsabaToth** [comment]
> It would be interesting to know if this problems only appear in the release builds
> If so then it's probably because Proguard strips out the code.

One more thing: ProGuard is now R8, and soon `android.enableR8=true` will be deprecated and cannot be turned off. I've seen workaround proposals which would exclude classes from ProGuard shrinking/removal, but since we are not dealing ProGuard any more those obviously won't help.

**Comment by MrCsabaToth** [comment]
> It would be interesting to know if this problems only appear in the release builds
> If so then it's probably because Proguard strips out the code.

Sir, you were right. This exactly explains why it's release only, and ProGuard rules (multiple) solved my particular problem.

**Comment by esDotDev** [comment]
From a developer perspective this is very frustrating as well. Any time something goes wrong in 1 plugin, all the plugins start throwing this error. If the packages are deeply integrated into your app, this is a bit of a debugging nightmare. You have no idea where to start, or what is causing the problem. 

We really need more clear errors here, pointing specifically to the plugin that has failed first and allowing it to expose it's errors. 

For example, currently I have this error coming from (at least) 4 plugins, only on iOS, on Android everything works perfectly, and all of these plugins support ios: 
* PathProvider
* FlutterLocalNotifications
* FirebaseCore
* Purchases_Flutter

No clue which of these actually has the error, and ripping any of them out of the app is very time consuming. This is exacerbated by long build times and test cycles on IOS. 

Very frustrating process and really needs more attention from the team. Pitching Flutter as production ready, but then having crippling issues like  this when you reach scale is a bit irresponsible.

**Comment by MrCsabaToth** [comment]
> From a developer perspective this is very frustrating as well. Any time something goes wrong in 1 plugin, all the plugins start throwing this error. If the packages are deeply integrated into your app, this is a bit of a debugging nightmare. You have no idea where to start, or what is causing the problem.
> 
> We really need more clear errors here, pointing specifically to the plugin that has failed first and allowing it to expose it's errors.
> 
> For example, currently I have this error coming from (at least) 4 plugins, only on iOS, on Android everything works perfectly, and all of these plugins support ios:
> 
>     * PathProvider
> 
>     * FlutterLocalNotifications
> 
>     * FirebaseCore
> 
>     * Purchases_Flutter
> 
> 
> No clue which of these actually has the error, and ripping any of them out of the app is very time consuming. This is exacerbated by long build times and test cycles on IOS.
> 
> Very frustrating process and really needs more attention from the team. Pitching Flutter as production ready, but then having crippling issues like this when you reach scale is a bit irresponsible.

Sorry to hear your struggle, and that iOS is affected as well.
Does iOS build pipeline has a similar stage as R8/ProGuard. I'm not iOS developer but that stack is different because ObjectiveC is more bare metal language, no VM like in case of Java/Kotlin so maybe ProGuard stage doesn't make sense that much? But maybe there's still some compiler stage which tries to remove unused code?

**Comment by esDotDev** [comment]
Really have no idea what's going on. If I take the same set of plugins to a fresh project on the same machine and simulator, it works fine. But if I do reduce my my view you a single Widget, that only prints `getApplicationSupportDirectory` it still throws this `MissingPluginException` error. 

I just toasted my entire iOS folder, and ran a new `create .` which I think might have got it working.

**Comment by febg11** [comment]
I have just installed my app using flutter build apk --release and I am getting this exception 

```
Unhandled Exception: MissingPluginException(No implementation found for method getTemporaryDirectory on channel plugins.flutter.io/path_provider)
```

It works fine when I use flutter run --release which is odd.

**Comment by escamoteur** [comment]
@febg11 If this is on Android, did you disable Proguard?

**Comment by febg11** [comment]
@escamoteur it is for Android and all I wrote was 

flutter build apk  

Do you disable proguard using the --no-shrink argument? I want to avoid using that if so.

**Comment by escamoteur** [comment]
Yep, the reason this makes problems is that proguarf stripps the plugin code out of the apk because it does not register correctly.

**Comment by febg11** [comment]
Makes sense. I did the android version migration yesterday and before that If I built the app without no shrink it would just crash completely when opened. 

Could I have made a mistake during the migration that would lead to this problem?

**Comment by escamoteur** [comment]
I don't think so. the Flutter Team changed the default for proguard somewhere in the recent passt, so probably you also made a Flutter Upgrade.

**Comment by febg11** [comment]
I remember running it.. i was updating the firebase dependencies. The actual error when thrown in the logcat comes from my code calling path provider and not a library, is that to be expected?

**Comment by r4mbo7** [comment]
I have followed the instructions here: https://flutter.dev/docs/cookbook/persistence/sqlite#example

And the latest command `flutter run lib/db_test.dart` returns this exception `Error: MissingPluginException(No implementation found for method getDatabasesPath on channel com.tekartik.sqflite)`.

I will come back and try again in few months. Good luck to the flutter team!

FTR
```
Flutter 1.24.0-10.2.pre • channel beta • https://github.com/flutter/flutter.git
Framework • revision 022b333a08 (7 weeks ago) • 2020-11-18 11:35:09 -0800
Engine • revision 07c1eed46b
Tools • Dart 2.12.0 (build 2.12.0-29.10.beta)
```

**Comment by escamoteur** [comment]
@r4mbo7 did you try to deactivate proguard? If this only happens in relase on Android that probably fixes it

**Comment by MrCsabaToth** [comment]
> @r4mbo7 did you try to deactivate proguard? If this only happens in relase on Android that probably fixes it

What's the best way to deactivate ProGuard from the build gradle? I'd prefer that over `--no-shrink`.

**Comment by escamoteur** [comment]
Please see here:
https://github.com/flutter/flutter/issues/72970#issuecomment-755336666

--no-shrink doesn't work I tried it

**Comment by MrCsabaToth** [comment]
> Please see here:
> [#72970 (comment)](https://github.com/flutter/flutter/issues/72970#issuecomment-755336666)
> 
> --no-shrink doesn't work I tried it

All what worked for me so far is a two part ProGuard exclusion rule, which is tailored to my use-case (flutter-blue plugin and file picker plugin fail), but I still see NPEs sporadically, so I'll try fully disable ProGuard. When I tried those switches I ran into compilation errors.

```
-keep class com.pauldemarco.flutter_blue.** { *; }
-keep class androidx.lifecycle.DefaultLifecycleObserver
```

**Comment by MrCsabaToth** [comment]
> Please see here:
> [#72970 (comment)](https://github.com/flutter/flutter/issues/72970#issuecomment-755336666)
> 
> --no-shrink doesn't work I tried it

Do you advise to try `android.enableR8=false` in `gradle.properties`? It's frightening to know that it'll go away soon with newer Gradle versions although we have this whole bucket of problems in our hand.

**Comment by escamoteur** [comment]
Sorry, I'm not an gradle specialist.

**Comment by gerryhigh** [comment]
> Sorry, I'm not an gradle specialist.

Are there any?

**Comment by jinfagang** [comment]
@pedromassango  I got same error in release mode:

```
lutter (18178): [ERROR:flutter/lib/ui/ui_dart_state.cc(186)] Unhandled Exception: MissingPluginException(No implementation found for method getAll on channel plugins.flutter.io/shared_preferences)

....

flutter (18178): [ERROR:flutter/lib/ui/ui_dart_state.cc(186)] Unhandled Exception: MissingPluginException(No implementation found for method getApplicationDocumentsDirectory on channel plugins.flutter.io/path_provider)
```

what should I do? I am using the latest flutter master branch!

**Comment by atrope** [comment]
@jinfagang 
Try this: https://github.com/FirebaseExtended/flutterfire/issues/4424#issuecomment-772750576

You will be able to use all your plugins if they use V2 Embedding

**Comment by ryanheise** [comment]
@atrope nice. You seem to have confirmed my theory from earlier up in the thread:

> Could it be that as the generated plugin registrant goes through each plugin in alphabetical order, if encounters one where `registerWithRegistry` fails, then all of the ones before it will be registered, but all of the ones after it will not?

Yes, usually the cause is an Android plugin that hasn't been upgraded from the v1 to the v2 architecture. But this error can also happen on iOS, in which case it has some other cause. And as others have mentioned above and elsewhere, the error can also occur during code shrinking, hot reloading with changes to plugins, and probably other reasons.

So until this error is completely eradicated, it is equally important as fixing it to actually improve the error message so that it doesn't lead users to go and post bug reports on the wrong plugins.

The audio_service project has 23 reported issues for MissingPluginException, and along with the other few plugins that create FlutterEngines, it finds itself at the epicentre of this bug. I ended up creating my own FAQ to help users track down the culprit plugin, so that one could decide whether to replace it with an alternative, report the bug to the correct author, or submit a pull request to the correct project.

But regarding the alphabetical order of the generated plugin registrant, I haven't tried your solution but if all it takes is one v1 plugin to halt the loading of all plugins after it, I suspect that this might only work to allow all of the v2 plugins to work within the newly spawned FlutterEngine, but the following v1 plugins would still not work within that newly spawned FlutterEngine. If your app by chance just happens to only use those v1 plugins in the main FlutterEngine, that might work out OK. Have you confirmed whether these v1 plugins operate correctly within the spawned FlutterEngine? If it actually does work, then this is a fix to the original v1 plugin registration problem for Android, and you might want to submit a pull request.

**Comment by atrope** [comment]
@ryanheise as I wrote there, it is a workaround to use V2 extensions on background services (Android) that should automatically register and any V1 extension was breaking the rest. 

I think V1 extensions need to be registered manually just like we did with firebase_messaging until version 7. 

So on Android we will be able to use any V2 extension automatically(with the workaround or if flutter team fixes the V1 breaking V2 plugin, or insert the workaround in the tool) and need to register Any V1 extension. 

Maybe the users on your app are trying to use a V1 extension that needs to manually be registered, do you know the names of the plugins so we can be sure it is this?

On iOS I have never tried using plugins in background.. ao I can't tell you right now, I will try to check and get back here.

**Comment by blasten** [comment]
@xster do you have any thoughts on this issue? Do you we have any tests that use background execution?

**Comment by MrCsabaToth** [comment]
> The audio_service project has 23 reported issues for MissingPluginException, and along with the other few plugins that create FlutterEngines, it finds itself at the epicentre of this bug. I ended up creating my own FAQ to help users track down the culprit plugin, so that one could decide whether to replace it with an alternative, report the bug to the correct author, or submit a pull request to the correct project.

I found your FAQ: https://github.com/ryanheise/audio_service/wiki/FAQ#why-does-my-app-get-a-missingpluginexception

**Comment by MrCsabaToth** [comment]
Just an idea for this latest V1/V2 conversation: if the suspicion turns out to be true then the best would be to display the V1/V2 compatibility status explicitly on pub.dev and possibly penalize V1 plugins through Pub Points. This could incentivize authors to upgrade to V2.

**Comment by rockingdice** [comment]
Please flutter team, do something to save the developers!

**Comment by rockingdice** [comment]
I made a repo for reproducing the issue, please follow the steps from this issue: #76768

**Comment by xster** [comment]
FWIW, I think a lot of cases are related to https://github.com/flutter/plugins/pull/3746 which was just merged as a fix. 

https://github.com/flutter/engine/pull/25297 will also add a bit more hints to the output when similar classes of issues arise. 

I'd like to see what impact these have in the next release.

**Comment by xster** [comment]
At the current stage, I'm broadly considering https://github.com/flutter/plugins/pull/3746 to have fixed all known issues and #78964 + https://github.com/flutter/engine/pull/25395 will make it much easier to detect future issues. 

I spot checked a few of the referenced bugs from here (thanks @jamesncl for compiling 
https://github.com/flutter/flutter/issues/62006#issuecomment-712244696) and I believe all have been resolved. If you're still seeing MissingPluginExceptions **on the master branch**, please open a new issue and cc me.

**Comment by stuartmorgan** [comment]
Leaving open for the initial request of having an FAQ about this exception, which would still be useful since there are still several ways of getting it.

**Comment by raj457036** [comment]
if anyone getting this in flutter 2.2.X

just update replace your contents of **MainActivity.kt** with

```kotlin
import androidx.annotation.NonNull;
import io.flutter.embedding.android.FlutterActivity
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugins.GeneratedPluginRegistrant

class MainActivity: FlutterActivity() {
    override fun configureFlutterEngine(@NonNull flutterEngine: FlutterEngine) {
        GeneratedPluginRegistrant.registerWith(flutterEngine);
    }
}

```

**Comment by stuartmorgan** [comment]
> if anyone getting this in flutter 2.2.X and Android
> 
> just

Again, there are a number of possible causes for a missing plugin exception. There is no one thing everyone who sees it (even on a single platform) can do that will fix it.

**Comment by MrCsabaToth** [comment]
> if anyone getting this in flutter 2.2.X and Android
> 
> just replace the contents of **MainActivity.kt** with
> 
> ```kotlin
> import androidx.annotation.NonNull;
> import io.flutter.embedding.android.FlutterActivity
> import io.flutter.embedding.engine.FlutterEngine
> import io.flutter.plugins.GeneratedPluginRegistrant
> 
> class MainActivity: FlutterActivity() {
>     override fun configureFlutterEngine(@NonNull flutterEngine: FlutterEngine) {
>         GeneratedPluginRegistrant.registerWith(flutterEngine);
>     }
> }
> ```

The `io.flutter.embedding.android.FlutterActivity` what you override here already has the exact same `configureFlutterEngine` implementation (source ocde in `io.flutter/flutter_embedding_debug`)! For this exact reason [the upgrade guide](https://github.com/flutter/flutter/wiki/Upgrading-pre-1.12-Android-projects) even advises to remove the unnecessarily empty file ("you can also optionally delete the MainActivity.java/kt file if you'd like"). Therefore I'm curious what would be the reason this could change anything? Maybe some old source which haven't been upgraded fully yet?

**Comment by wanjm** [comment]
is there any update for it, I come across many plugins that missing; any work around for all; I don't like --no-shrink;
MissingPluginException(No implementation found for method toggle on channel wakelock);

**Comment by stuartmorgan** [comment]
I'm going to lock this issue as it's continuing to attract comments that are trying to use this a single location to debug something that has, as discussed in recent comments above, many possible unrelated causes. If you are trying to report an issue where you believe a MissingPluginException is due to a Flutter bug, please **file a new issue for your specific case**.

This issue is at this point only tracking *adding documentation* about possible sources of this issue, not a place to report issues themselves.
